src/Data/Bin/Properties.agda

```diff
@@ -1,143 +1,867 @@
-open import Relation.Nullary using (yes; no)
+open import Algebra
+import      Algebra.FunctionProperties as FuncProp
+open import Algebra.FunctionProperties.Consequences.Propositional using
+                            (comm+cancelˡ⇒cancelʳ; comm+distrˡ⇒distrʳ)
+open import Algebra.Structures
+open import Data.Bin.Base
+open import Data.Nat using (ℕ; s≤s)
+                     renaming (suc to 1+_; pred to predₙ; _+_ to _+ₙ_;
+                               _*_ to _*ₙ_; _∸_ to _∸ₙ_; _≤_ to _≤ₙ_)
 
-------------------------------------------------------------------------
--- (Bin, _≡_) is a decidable setoid
-
-1#-injective : ∀ {as bs} → as 1# ≡ bs 1# → as ≡ bs
-1#-injective refl = refl
-
-infix 4 _≟_ _≟ₑ_
-
-_≟ₑ_ : ∀ {base} → Decidable (_≡_ {A = Expansion base})
-_≟ₑ_ []       []       = yes refl
-_≟ₑ_ []       (_ ∷ _)  = no λ()
-_≟ₑ_ (_ ∷ _) []        = no λ()
-_≟ₑ_ (x ∷ xs) (y ∷ ys) with x Fin.≟ y | xs ≟ₑ ys
-... | _        | no xs≢ys = no (xs≢ys ∘ proj₂ ∘ ∷-injective)
-... | no  x≢y  | _        = no (x≢y   ∘ proj₁ ∘ ∷-injective)
-... | yes refl | yes refl = yes refl
-
-_≟_ : Decidable {A = Bin} _≡_
-0#    ≟ 0#    = yes refl
-0#    ≟ bs 1# = no λ()
-as 1# ≟ 0#    = no λ()
-as 1# ≟ bs 1# with as ≟ₑ bs
-... | yes refl  = yes refl
-... | no  as≢bs = no (as≢bs ∘ 1#-injective)
-
-≡-isDecEquivalence : IsDecEquivalence _≡_
-≡-isDecEquivalence = record
+open import Data.Nat.Properties as ℕp using (even≢odd)
+open import Data.Product    using (_,_; proj₁; proj₂; ∃)
+open import Data.Sum        using (_⊎_; inj₁; inj₂)
+open import Function        using (_∘_; _$_; id; const; case_of_)
+open import Level           using () renaming (zero to 0ℓ)
+open import Relation.Binary using (Setoid; DecSetoid; Decidable;
+                                                      IsDecEquivalence)
+open import Relation.Binary.PropositionalEquality as P using
+     (_≡_; _≢_; _≗_; refl; sym; trans; cong; cong₂; subst; isEquivalence)
+open import Relation.Nullary          using (yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+
+import Algebra.Properties.CommutativeSemigroup ℕp.+-semigroup ℕp.+-comm
+                                                     as Of+ℕ-semigroup
+import Algebra.Properties.CommutativeSemigroup ℕp.*-semigroup ℕp.*-comm
+                                                     as Of*ℕ-semigroup
+open FuncProp (_≡_ {A = Bin})
+open P.≡-Reasoning
+open Bin
+
+
+------------------------------------------------------------------------------
+-- Equality properties
+------------------------------------------------------------------------------
+
+≡-setoid : Setoid 0ℓ 0ℓ
+≡-setoid = P.setoid Bin
+
+2suc≢0 :  ∀ {x} → 2suc x ≢ 0#
+2suc≢0 ()
+
+suc2*≢0 :  ∀ {x} → suc2* x ≢ 0#
+suc2*≢0 ()
+
+2suc-injective : ∀ {x y} → 2suc x ≡ 2suc y → x ≡ y
+2suc-injective {0#}      {0#}      _    =  refl
+2suc-injective {2suc _}  {2suc _}  refl =  refl
+2suc-injective {suc2* _} {suc2* _} refl =  refl
+
```

Comment:
`2suc` is a constructor. As such it is always injective.
You should be able to write `2suc-injective refl =  refl`.

Reply:
Done